scheduler/scheduler.go

```diff
@@ -1,122 +1,328 @@
+type Options struct {
+
+	// MetricsUpdateTimeout defines the frequence of how often the LIFO metrics
+	// are updated when they are enabled. Defaults to 1s.
+	MetricsUpdateTimeout time.Duration
+
+	// EnableRouteLIFOMetrics enables collecting metrics about the LIFO queues.
+	EnableRouteLIFOMetrics bool
+
+	// Metrics must be provided to the registry in order to collect the LIFO metrics.
+	Metrics metrics.Metrics
+}
+
+// Registry maintains a set of LIFO queues. It is used to preserve LIFO queue instances
+// across multiple generations of the routing. It implements the routing.PostProcessor
+// interface, it is enough to just pass in to routing.Routing when initializing it.
 type Registry struct {
-	mu          sync.Mutex
-	groupConfig map[string]Config
-	stacks      map[string]*Stack
+	options   Options
+	queues    *sync.Map
+	measuring bool
+	quit      chan struct{}
 }
 
+// LIFOFilter is the interface that needs to be implemented by the filters that
+// use a LIFO queue maintained by the registry.
 type LIFOFilter interface {
-	SetStack(*Stack)
-	GetStack() *Stack
-	Config(*Registry) Config
-	Key() string
-	SetKey(string)
+
+	// SetQueue will be used by the registry to pass in the right queue to
+	// the filter.
+	SetQueue(*Queue)
+
+	// GetQueue is currently used only by tests.
+	GetQueue() *Queue
+
+	// Config will be called by the registry once during processing the
+	// routing to get the right queue settings from the filter.
+	Config() Config
 }
 
-func newStack(c Config) *Stack {
-	return &Stack{
-		config: c,
-		stack: jobqueue.With(jobqueue.Options{
-			MaxConcurrency: c.MaxConcurrency,
-			MaxStackSize:   c.MaxStackSize,
-			Timeout:        c.Timeout,
-		}),
+// GroupedLIFOFilter is an extension of the LIFOFilter interface for filters
+// that use a shared queue.
+type GroupedLIFOFilter interface {
+	LIFOFilter
+
+	// Group returns the name of the group.
+	Group() string
+
+	// HasConfig indicates that the current filter provides the queue
+	// queue settings for the group.
+	HasConfig() bool
+}
+
+// Wait blocks until a request can be processed or needs to be rejected.
+// When it can be processed, calling done indicates that it has finished.
+// It is mandatory to call done() the request was processed. When the
+// request needs to be rejected, an error will be returned.
+func (q *Queue) Wait() (done func(), err error) {
+	return q.queue.Wait()
+}
+
+// Status returns the current status of a queue.
+func (q *Queue) Status() QueueStatus {
+	st := q.queue.Status()
+	return QueueStatus{
+		ActiveRequests: st.ActiveJobs,
+		QueuedRequests: st.QueuedJobs,
 	}
 }
 
-func (s *Stack) Wait() (done func(), err error) {
-	return s.stack.Wait()
+// Config returns the configuration that the queue was created with.
+func (q *Queue) Config() Config {
+	return q.config
 }
 
-func (s *Stack) close() {
-	s.stack.Close()
+func (q *Queue) reconfigure() {
+	// renaming Stack -> Queue in the jobqueue project will follow
```

Comment:
comment is not clear to me
Does it mean that a rename would be a recreate (I suppose it would be)?

Reply:
I'm removing the comment. It's simply that in the jobqueue package, while the project was renamed from jobstack to jobqueue, the exported structure is still called Stack. It should be Queue.

Reply:
ok